text/0000-const-trait-impls.md

```diff
@@ -0,0 +1,902 @@
+}
+const fn bar<T: ~const Destruct>(t: T) {
+    // Fine, we can safely invoke the destructor of `T`.
+}
+```
+
+When a value of a generic type goes out of scope, it is dropped and (if it has one) its `Drop` impl gets invoked.
+This situation seems no different from other trait bounds, except that types can be dropped without implementing `Drop`
+(as they can contain types that implement `Drop`). In that case the type's drop glue is invoked.
+
+The `Destruct` trait is a bound for whether a type has drop glue. This is trivally true for all types.
+
+`~const Destruct` trait bounds are satisfied only if the type's `Drop` impl (if any) is `const` and all of the types of
+its components are `~const Destruct`.
+
+While this means that it's a breaking change to add a type with a non-const `Drop` impl to a type,
+that's already true and nothing new:
+
+```rust
+pub struct S {
+    x: u8,
+    y: Box<()>, // adding this field breaks code.
+}
+
+const fn f(_: S) {}
+//~^ ERROR destructor of `S` cannot be evaluated at compile-time
+```
+
+## Trivially enabled features
+
+You can use `==` operators on most types from libstd from within const contexts.
+
+```rust
+const _: () = {
+    let a = [1, 2, 3];
+    let b = [4, 5, 6];
+    assert!(a != b);
+};
+const _: () = {
+    let a = Some(42);
+    let b = a;
+    assert!(a == b);
+};
+```
+
+Note that the use of `assert_eq!` is waiting on `Debug` impls becoming `const`, which
+will likely be tracked under a separate feature gate under the purview of T-libs.
+Similarly other traits will be made `const` over time, but doing so will be
+unblocked by this feature.
+
+## Crate authors: Making your own custom types easier to use
+
+You can write const trait impls of many standard library traits for your own types.
+While it was often possible to write the same code in inherent methods, operators were
+covered by traits from `std::ops` and thus not avaiable for const contexts.
+Most of the time it suffices to add `const` before the trait name in the impl block.
+The compiler will guide you and suggest where to also
+add `~const` bounds for trait bounds on generic parameters of methods or the impl.
+
+Similarly you can make your traits available for users of your crate to implement constly.
+Note that this will change your semver guarantees: you are now guaranteeing that any future
+methods you add don't just have a default body, but a `const` default body. The compiler will
+enforce this, so you can't accidentally make a mistake, but it may still limit how you can
+extend your trait without having to do a major version bump.
+Most of the time it suffices to add `const` before the `trait` declaration. The compiler will
+guide you and suggest where to also add `~const` bounds for super trait bounds or trait bounds
+on generic parameters of your trait or your methods.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## How does this work in the compiler?
```

Comment:
I got entirely lost in this part. And that is despite the fact that I heard of the "host " effect before, which definitely needs more introduction: it means "code needs access to the host machine" aka "not const". I get the desire for not using negations in the name, but "not-const" would be a lot more intuitive I think...

Reply:
I think runtime would be a more intuitive name.

Reply:
I just renamed it to `constness` in order to not introduce new terminology, but gave a one sentence explainer that the `host` name of the internal data structure is here due to our impl history. We can keep it, as the internal docs for it explain it, or we can change it again in the future now that we don't use a const generic bool anymore but just have an enum with names for things